server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropLoader.java

```diff
@@ -0,0 +1,169 @@
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.accumulo.fate.zookeeper.ZooReaderWriter;
+import org.apache.accumulo.server.conf.codec.VersionedPropCodec;
+import org.apache.accumulo.server.conf.codec.VersionedProperties;
+import org.apache.accumulo.server.conf.store.PropCacheId;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.data.Stat;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.benmanes.caffeine.cache.CacheLoader;
+
+public class ZooPropLoader implements CacheLoader<PropCacheId,VersionedProperties> {
+
+  private static final Logger log = LoggerFactory.getLogger(ZooPropLoader.class);
+
+  private final ZooReaderWriter zrw;
+  private final VersionedPropCodec propCodec;
+  // used to set watcher, does not react to events.
+  private final PropStoreWatcher propStoreWatcher;
+  private final PropStoreMetrics metrics;
+
+  public ZooPropLoader(final ZooReaderWriter zrw, final VersionedPropCodec propCodec,
+      final PropStoreWatcher propStoreWatcher, final PropStoreMetrics metrics) {
+    this.zrw = zrw;
+    this.propCodec = propCodec;
+    this.propStoreWatcher = propStoreWatcher;
+    this.metrics = metrics;
+  }
+
+  @Override
+  public @Nullable VersionedProperties load(PropCacheId propCacheId) {
+    try {
+      log.trace("load called for {}", propCacheId);
+      Stat stat = new Stat();
+      long startNanos = System.nanoTime();
+
+      var vProps = propCodec.fromBytes(zrw.getData(propCacheId.getPath(), propStoreWatcher, stat));
+
+      metrics.addLoadTime(
+          TimeUnit.MILLISECONDS.convert(System.nanoTime() - startNanos, TimeUnit.NANOSECONDS));
+
+      if (stat.getVersion() != vProps.getDataVersion()) {
+        log.warn("data versions between decoded value {} and zk value {} do not match",
+            vProps.getDataVersion(), stat.getVersion());
+        return null;
+      }
+      return vProps;
+    } catch (Exception ex) {
+      metrics.incrZkError();
+      log.warn("Failed to load: {} from ZooKeeper, returning null", propCacheId, ex);
+      propStoreWatcher.signalZkChangeEvent(propCacheId);
+      return null;
+    }
+  }
+
+  @Override
+  public CompletableFuture<? extends VersionedProperties> asyncLoad(PropCacheId propCacheId,
+      Executor executor) {
+    log.trace("asyncLoad called for key: {}", propCacheId);
+    return CacheLoader.super.asyncLoad(propCacheId, executor);
+  }
+
+  @Override
+  public CompletableFuture<? extends VersionedProperties> asyncReload(PropCacheId propCacheId,
+      VersionedProperties oldValue, Executor executor) throws Exception {
+    log.trace("asyncReload called for key: {}", propCacheId);
+    metrics.incrRefresh();
+
+    return CompletableFuture.supplyAsync(() -> loadIfDifferentVersion(propCacheId, oldValue));
```

Comment:
`supplyAsync` may use a jvm wide shared executor service.  Seems like this can cause unrelated code in the JVM to interfere with each other, but I am not sure about this.

Reply:
ZooPropLoader is supplied to the Caffine cache and run in the context of the cache code.  There is a thread pool specified there and I think Caffeine will use that pool and not the general JVM pool.

Reply:
> There is a thread pool specified there and I think Caffeine will use that pool and not the general JVM pool.

CompletableFuture.supplyAsync() is a java function that has no awareness of any caffeine thread pools.  Looking at the [javadoc](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/CompletableFuture.html#supplyAsync(java.util.function.Supplier)) it says.

> Returns a new CompletableFuture that is asynchronously completed by a task running in the ForkJoinPool.commonPool() with the value obtained by calling the given Supplier.

That ForkJoinPool.commonPool() is static JVM wide thread pool.

There is another version of [supply async that takes a thread pool](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/CompletableFuture.html#supplyAsync(java.util.function.Supplier,java.util.concurrent.Executor)).

Reply:
Great catch - the executor from the cache was being passed in to the function - I just neglected to use it here.

Reply:
Addressed in e8e60988